internal/staging/primitives/inclusionproof.go

```diff
@@ -0,0 +1,207 @@
+// Copyright (c) 2019-2021 The Decred developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package primitives
+
+import (
+	"github.com/decred/dcrd/chaincfg/chainhash"
+)
+
+// log2FloorMasks defines the masks to use when quickly calculating
+// floor(log2(x)) in a constant log2(32) = 5 steps, where x is a uint32, using
+// shifts.  They are derived from (2^(2^x) - 1) * (2^(2^x)), for x in 4..0.
+var log2FloorMasks = []uint32{0xffff0000, 0xff00, 0xf0, 0xc, 0x2}
+
+// fastLog2Ceil calculates and returns ceil(log2(x)) in a constant 5 steps.
+func fastLog2Ceil(n uint32) uint8 {
+	// The final answer will be one more when the input is greater than one and
+	// not a power of two.
+	rv := uint8(0)
+	if n&(n-1) != 0 {
+		rv++
+	}
+	exponent := uint8(16)
+	for i := 0; i < 5; i++ {
+		if n&log2FloorMasks[i] != 0 {
+			rv += exponent
+			n >>= exponent
+		}
+		exponent >>= 1
+	}
+	return rv
+}
+
+// GenerateInclusionProof treats the provided slice of hashes as leaves of a
+// merkle tree and generates and returns a merkle tree inclusion proof for the
+// given leaf index.  The proof can be used to efficiently prove the leaf
+// associated with given leaf index is a member of the tree.
+//
+// A merkle tree inclusion proof consists of the ceil(log2(x)) intermediate
+// sibling hashes along the path from the target leaf to prove through the root
+// node.  The sibling hashes, along with the original leaf hash (and its
+// original leaf index), can be used to recalculate the merkle root which, in
+// turn, can be verified against a known good merkle root in order to prove the
+// leaf is actually a member of the tree at that position.
+//
+// For example, consider the following merkle tree:
+//
+//	         root = h1234 = h(h12 + h34)
+//	        /                           \
+//	  h12 = h(h1 + h2)            h34 = h(h3 + h4)
+//	   /            \              /            \
+//	  h1            h2            h3            h4
+//
+// Further, consider the goal is to prove inclusion of h3 at the 0-based leaf
+// index of 2.  The proof will consist of the sibling hashes h4 and h12.  On the
+// other hand, if the goal were to prove inclusion of h2 at the 0-based leaf
+// index of 1, the proof would consist of the sibling hashes h1 and h34.
+//
+// Specifying a leaf index that is out of range will return nil.
+func GenerateInclusionProof(leaves []chainhash.Hash, leafIndex uint32) []chainhash.Hash {
+	if len(leaves) == 0 {
+		return nil
+	}
+
```

Comment:
The diagram says to return the **sibling** hash list.  A single leaf "tree" only consists of leaf index 0 and therefore there is no sibling hash at index 1 which means it results in an empty list.  Note that the Merkle root of a single leaf "tree" is just that leaf hash.

You can convince yourself of that by working through the code or just trying it and watch the tests fail.  First, it would fail because the generated proof wouldn't match the expected it, and if you were to make the proof match the incorrect value, it would then fail to verify.

That said, we could certainly short circuit it here since the result is necessarily going to be an empty list, but I'm not sure it's really worth changing the code since it's never called with a single leaf in practice given everything currently already understands that the proof is going to going to be empty and just does the short circuiting itself.

For example, the code that serves a v2 gcs filter with a proof hash:

```Go
	// NOTE: When more header commitments are added, this will need to load the
	// inclusion proof for the filter from the database.  However, since there
	// is only currently a single commitment, there is only a single leaf in the
	// commitment merkle tree, and hence the proof hashes will always be empty
	// given there are no siblings.  Adding an additional header commitment will
	// require a consensus vote anyway and this can be updated at that time.
```

Reply:
I went ahead and changed this to short circuit since the check is effectively the same cost (less than versus equality) and this module will ultimately be available to all consumers, even those outside of dcrd who might not be so focused on performance that they avoid calling it in that case.

Reply:
Thank you for the explanation.